cache/fifo_test.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package cache
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestFIFOCache(t *testing.T) {
+	tests := []struct {
+		name     string
+		limit    int
+		maxVal   int
+		expected error
+		check    func(t *testing.T, cache *FIFO[int, int])
+	}{
+		{
+			name:     "empty cache fails",
+			limit:    0,
+			expected: errors.New("maxSize must be greater than 0"),
+		},
+		{
+			name:   "inserting up to limit works",
+			limit:  10,
+			maxVal: 9,
+		},
+		{
+			name:   "inserting after limit cleans first",
+			limit:  10,
+			maxVal: 10,
+			check: func(t *testing.T, cache *FIFO[int, int]) {
+				require := require.New(t)
+				_, ok := cache.Get(0)
+				require.False(ok)
+				v, ok := cache.Get(1)
+				require.True(ok)
+				require.Equal(1, v)
+			},
+		},
+		{
+			name:   "inserting before limit can insert again",
+			limit:  10,
+			maxVal: 8,
+			check: func(t *testing.T, cache *FIFO[int, int]) {
+				require := require.New(t)
+				exists := cache.Put(9, 9)
+				require.False(exists)
+				exists = cache.Put(10, 10)
+				require.False(exists)
+				_, ok := cache.Get(0)
+				require.False(ok)
+			},
+		},
+		{
+			name:   "inserting existing value doesn't free value",
+			limit:  10,
+			maxVal: 9,
+			check: func(t *testing.T, cache *FIFO[int, int]) {
+				require := require.New(t)
+				v, ok := cache.Get(0)
+				require.True(ok)
+				require.Equal(0, v)
+				exists := cache.Put(0, 10)
+				require.True(exists)
+				v, ok = cache.Get(0)
+				require.True(ok)
+				require.Equal(10, v)
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			require := require.New(t)
+
+			cache, err := NewFIFO[int, int](tt.limit)
+			require.Equal(tt.expected, err)
```

Comment:
nit: use `ErrorIs`. This does a match to check if the error was wrapped instead of just the text which is more durable ([ref](https://go.dev/blog/go1.13-errors))

Reply:
I can't use `ErrorIs`, the error that says "maxSize must be greater than 0" is private and cannot be imported. It's defined in https://github.com/ava-labs/avalanchego/blob/321e727328a86bcfdb8f5bef164e7d50e698408b/utils/buffer/bounded_nonblocking_queue.go#L11
This error check has been moved to a unit test `TestEmptyCacheFails` that checks that this error occurs when the cache is initialized with a length of 0, so it's better contained.